### PR TITLE
Fix warning header: include agent as host

### PR DESCRIPTION
**Pull request by pengchen0692**
## Description
Fix warning header: include agent as host

## Related issues
Addresses [AB#91168](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/91168)

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Api/DicomApiResource.resx [hunk 1]
```diff
@@ -173,7 +173,7 @@
     <comment>{0} is field.</comment>
   </data>
   <data name="WarningHeader" xml:space="preserve">
-    <value>299: {0}</value>
-    <comment>{0} is the warning message</comment>
+    <value>299 {0}: "{1}"</value>
+    <comment>{0} is host name, {1} is the warning message</comment>
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Api/DicomApiResource.resx:177]: This probably shouldn't be in here, as this `string` is not culture-specific.

File: src/Microsoft.Health.Dicom.Api/Extensions/HttpResponseExtensions.cs [hunk 2]
```diff
@@ -25,6 +25,19 @@ public static void AddLocationHeader(this HttpResponse response, Uri locationUrl
...
+    /// Set Response Warning header.
+    /// </summary>
+    /// <param name="response">The httpResponse.</param>
+    /// <param name="host">Host name.</param>
+    /// <param name="message">The warning message.</param>
+    public static void SetWarning(this HttpResponse response, string host, string message)
+    {
+        EnsureArg.IsNotNull(response, nameof(response));
+        EnsureArg.IsNotEmptyOrWhiteSpace(message, nameof(message));
+        response.Headers.Warning = string.Format(DicomApiResource.WarningHeader, host ?? string.Empty, message);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Api/Extensions/HttpResponseExtensions.cs:38]: If `host` is `null` or empty I think it needs to be `'-'` instead

Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.Api/Extensions/HttpResponseExtensions.cs]: yeah, that's right!

File: src/Microsoft.Health.Dicom.Api/DicomApiResource.resx (hunk 1 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Api/DicomApiResource.resx:177]: While no longer in use, this change still needs to be deleted

**Review by jovinson-ms — APPROVED** [review]

File: src/Microsoft.Health.Dicom.Core/Models/HttpWarningCode.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+namespace Microsoft.Health.Dicom.Core.Models;
+
+/// <summary>
+/// Represents Http Warning code.  https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Warning#warning_codes
+/// </summary>
+[System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1008:Enums should have zero value", Justification = "0 is not valid HttpWarningCode")]
+public enum HttpWarningCode
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Models/HttpWarningCode.cs:12]: Do we need this class? Why can't we direcly set the value from the controller

File: src/Microsoft.Health.Dicom.Api/Controllers/WorkitemController.Cancel.cs [hunk 4]
```diff
@@ -63,9 +63,16 @@ public async Task<IActionResult> CancelAsync(string workitemInstanceUid)
         if (response.Status is Core.Messages.Workitem.WorkitemResponseStatus.Conflict
             && !string.IsNullOrEmpty(response.Message))
         {
-            Response.Headers.Warning = string.Format(DicomApiResource.WarningHeader, response.Message);
+            Response.SetWarning(Core.Models.HttpWarningCode.MiscPersistentWarning, GetHost(), response.Message);
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Api/Controllers/WorkitemController.Cancel.cs:66]: Why can't we directly set the code, why do we need the enum?

File: src/Microsoft.Health.Dicom.Api/DicomApiResource.resx (hunk 1 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.Api/DicomApiResource.resx:177]: yeah, forgot to do that...

File: src/Microsoft.Health.Dicom.Api/Controllers/WorkitemController.Cancel.cs [hunk 5]
```diff
@@ -63,9 +63,16 @@ public async Task<IActionResult> CancelAsync(string workitemInstanceUid)
...
         }
 
         return StatusCode((int)response.Status.CancelResponseToHttpStatusCode(), response.Message);
     }
+
+    private string GetHost()
+    {
+        string host = Request.Host.Host;
+        // As Dicom standard, should append colon after service. https://dicom.nema.org/medical/dicom/current/output/chtml/part18/sect_11.7.3.2.html
+        return string.IsNullOrWhiteSpace(host) ? host : host + ":";
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Api/Controllers/WorkitemController.Cancel.cs:76]: Is the colon still present if the host is missing and replaced with `'-'`?

**Review by wsugarman — APPROVED** [review]

File: src/Microsoft.Health.Dicom.Api/Controllers/WorkitemController.Cancel.cs (hunk 5 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.Api/Controllers/WorkitemController.Cancel.cs:76]: yeah, that's the trickly part.  As http standard, it should be `'_'`.

**Pull request merged**